### PR TITLE
Fix: after BIND_TABLE no columns in object

### PR DESCRIPTION
**Pull request by RedWolf112**

### CODE REVIEW

**Comment by RedWolf112** [comment]
After method bind_table in the worksheet the columns are empty

**Comment by gregorwolf** [comment]
Hi @RedWolf112 ,
in which example report can I test the improved behaviour?

**Comment by RedWolf112** [comment]
Hi @gregorwolf,

at the moment ther is no test report.
I have to write a new one. 
The test report should use the get_columns_iterator( ) from the worksheet.
Without the fix the columns iterator has no entries.

**Comment by AndreaBorgia-Abo** [comment]
#781 was merged and resulted in some conflicts, sorry for the inconvenience.

**Comment by sandraros** [comment]
There is the same issue even after `SET_CELL`, the column collection is empty, and the rows iterator is also concerned.

Here is a test class with 8 methods which fail, to help support the fix:
```
REPORT.
CLASS ltc_main DEFINITION
      FOR TESTING
      DURATION SHORT
      RISK LEVEL HARMLESS.
  PRIVATE SECTION.
    METHODS setup.
    METHODS get_count
      IMPORTING
        iterator      TYPE REF TO cl_object_collection_iterator
      RETURNING
        VALUE(result) TYPE i.
    METHODS set_cell_columns_iterator FOR TESTING.
    METHODS set_cell_rows_iterator FOR TESTING.
    METHODS bind_table_columns_iterator FOR TESTING.
    METHODS bind_table_rows_iterator FOR TESTING.
    METHODS set_cell_columns FOR TESTING.
    METHODS set_cell_rows FOR TESTING.
    METHODS bind_table_columns FOR TESTING.
    METHODS bind_table_rows FOR TESTING.
    DATA: lo_excel         TYPE REF TO zcl_excel,
          lo_worksheet     TYPE REF TO zcl_excel_worksheet,
          scarr_lines      TYPE STANDARD TABLE OF scarr,
          columns_iterator TYPE REF TO cl_object_collection_iterator,
          rows_iterator    TYPE REF TO cl_object_collection_iterator,
          columns TYPE REF TO zcl_excel_columns,
          rows    TYPE REF TO zcl_excel_rows.
ENDCLASS.
CLASS ltc_main IMPLEMENTATION.
  METHOD setup.
    SELECT * FROM scarr INTO TABLE @scarr_lines.
    lo_excel = NEW zcl_excel( ).
    lo_worksheet = lo_excel->get_active_worksheet( ).
  ENDMETHOD.
  METHOD set_cell_columns_iterator.
    lo_worksheet->set_cell( ip_column = 2 ip_row = 3 ip_value = 'a' ).
    columns_iterator = lo_worksheet->get_columns_iterator( ).
    cl_abap_unit_assert=>assert_true( columns_iterator->has_next( ) ).
    cl_abap_unit_assert=>assert_equals( act = get_count( columns_iterator ) exp = 1 ).
  ENDMETHOD.
  METHOD set_cell_rows_iterator.
    lo_worksheet->set_cell( ip_column = 2 ip_row = 3 ip_value = 'a' ).
    rows_iterator = lo_worksheet->get_rows_iterator( ).
    cl_abap_unit_assert=>assert_true( rows_iterator->has_next( ) ).
    cl_abap_unit_assert=>assert_equals( act = get_count( rows_iterator ) exp = 1 ).
  ENDMETHOD.
  METHOD bind_table_columns_iterator.
    lo_worksheet->bind_table(
        ip_table          = scarr_lines
        is_table_settings = VALUE zexcel_s_table_settings( top_left_column = 'B' top_left_row = 3 ) ).
    columns_iterator = lo_worksheet->get_columns_iterator( ).
    cl_abap_unit_assert=>assert_true( columns_iterator->has_next( ) ).
    cl_abap_unit_assert=>assert_equals( act = get_count( columns_iterator ) exp = 4 ).
  ENDMETHOD.
  METHOD bind_table_rows_iterator.
    lo_worksheet->bind_table(
        ip_table          = scarr_lines
        is_table_settings = VALUE zexcel_s_table_settings( top_left_column = 'B' top_left_row = 3 ) ).
    rows_iterator = lo_worksheet->get_rows_iterator( ).
    cl_abap_unit_assert=>assert_true( rows_iterator->has_next( ) ).
    cl_abap_unit_assert=>assert_equals( act = get_count( rows_iterator ) exp = lines( scarr_lines ) + 1 ).
  ENDMETHOD.
  METHOD set_cell_columns.
    lo_worksheet->set_cell( ip_column = 2 ip_row = 3 ip_value = 'a' ).
    columns = lo_worksheet->get_columns( ).
    cl_abap_unit_assert=>assert_equals( act = columns->size( ) exp = 1 ).
  ENDMETHOD.
  METHOD set_cell_rows.
    lo_worksheet->set_cell( ip_column = 2 ip_row = 3 ip_value = 'a' ).
    rows = lo_worksheet->get_rows( ).
    cl_abap_unit_assert=>assert_equals( act = rows->size( ) exp = 1 ).
  ENDMETHOD.
  METHOD bind_table_columns.
    lo_worksheet->bind_table(
        ip_table          = scarr_lines
        is_table_settings = VALUE zexcel_s_table_settings( top_left_column = 'B' top_left_row = 3 ) ).
    columns = lo_worksheet->get_columns( ).
    cl_abap_unit_assert=>assert_equals( act = columns->size( ) exp = 4 ).
  ENDMETHOD.
  METHOD bind_table_rows.
    lo_worksheet->bind_table(
        ip_table          = scarr_lines
        is_table_settings = VALUE zexcel_s_table_settings( top_left_column = 'B' top_left_row = 3 ) ).
    rows = lo_worksheet->get_rows( ).
    cl_abap_unit_assert=>assert_equals( act = rows->size( ) exp = lines( scarr_lines ) + 1 ).
  ENDMETHOD.
  METHOD get_count.
    result = 0.
    WHILE iterator->has_next( ) = abap_true.
      iterator->get_next( ).
      result = result + 1.
    ENDWHILE.
  ENDMETHOD.
ENDCLASS.

DATA: gc_save_file_name TYPE string VALUE 'PR_731.xlsx'.
INCLUDE zdemo_excel_outputopt_incl.
PARAMETERS iterator AS CHECKBOX.

START-OF-SELECTION.
  SELECT * FROM scarr INTO TABLE @DATA(scarr_lines).
  DATA(lo_excel) = NEW zcl_excel( ).
  DATA(lo_worksheet) = lo_excel->get_active_worksheet( ).
  lo_worksheet->set_cell( ip_column = 1 ip_row = 1 ip_value = 'a' ).
  lo_worksheet->bind_table(
      ip_table          = scarr_lines
      is_table_settings = VALUE zexcel_s_table_settings( top_left_column = 'A' top_left_row = 3 ) ).
  IF iterator = abap_true.
    DATA(lo_rows_iterator) = lo_worksheet->get_rows_iterator( ).
    WHILE abap_true = lo_rows_iterator->has_next( ).
      DATA(lo_row) = CAST zcl_excel_row( lo_rows_iterator->get_next( ) ).
      DATA(lo_columns_iterator) = lo_worksheet->get_columns_iterator( ).
      WHILE abap_true = lo_columns_iterator->has_next( ).
        DATA(lo_column) = CAST zcl_excel_column( lo_columns_iterator->get_next( ) ).
        lo_worksheet->get_cell(
          EXPORTING
            ip_column  = lo_column->get_column_index( )
            ip_row     = lo_row->get_row_index( )
          IMPORTING
            ep_value   = DATA(lv_value) ).
        WRITE: / 'row', lo_row->get_row_index( ), ', column:', lo_column->get_column_index( ), ', cell:', lv_value.
      ENDWHILE.
    ENDWHILE.
  ELSE.
    lcl_output=>output( lo_excel ).
  ENDIF.
```

Maybe we could complete the list of columns/rows when the iterator is instantiated?

It should be also the same for methods `GET_COLUMNS` and `GET_ROWS`.

For information, to know how many rows and columns are in the worksheet, the method `GET_DIMENSION_RANGE` can be used.

**Comment by sandraros** [comment]
@RedWolf112 @AndreaBorgia-Abo @gregorwolf I investigated more about the columns and rows iterators, and I think the original and current feature is to propose an iterator only for those columns and rows explicitly "added" using `sheet->get_column( 1 )` or `sheet->add_new_column( 1 )`/`sheet->get_row( 1 )` or `sheet->add_new_row( 1 )`, to define column and row attributes, and not to support the cells defined using `SET_CELL`, `BIND_TABLE` and so on. I would prefer a completely different solution from the one proposed here (maybe by calling a new method to explicitly synchronize columns and rows from the cells of the worksheet `sheet_content` attribute). I propose to create an issue to agree first on a solution, and to abandon this PR. Please tell me if you agree to go this way?

**Comment by AndreaBorgia-Abo** [comment]
I have no objections but @RedWolf112 [might](https://github.com/sapmentors/abap2xlsx/pull/731#issuecomment-777378478)

**Comment by RedWolf112** [comment]
I think it's a good solution

**Comment by sandraros** [comment]
@RedWolf112 @AndreaBorgia-Abo I pushed a new commit/new solution in the original branch. I extended the solution to also handle the methods `GET_ROWS` and `GET_COLUMNS`. I used the updated code snippet above to check the results (test class + output through iterator) but of course it would make more sense if you do other tests. Thank you for reviewing and merging (Andrea) if you agree!

**Comment by AndreaBorgia-Abo** [comment]
On it... tomorrow morning I be able to test and merge.

File: src/zcl_excel_worksheet.clas.abap [hunk 1]
```diff
@@ -2426,12 +2442,52 @@ CLASS zcl_excel_worksheet IMPLEMENTATION.
 
 
   METHOD get_rows.
+
+    DATA: first_row TYPE i,
+          last_row  TYPE i,
+          row       TYPE i.
```
Comment (AndreaBorgia-Abo) [inline src/zcl_excel_worksheet.clas.abap]:
```suggestion
    DATA: row       TYPE i.
```

File: src/zcl_excel_worksheet.clas.abap [hunk 2]
```diff
@@ -2426,12 +2442,52 @@ CLASS zcl_excel_worksheet IMPLEMENTATION.
...
+
+      first_row = rows->get_min_index( ).
+      IF first_row = 0.
+        first_row = zcl_excel_common=>c_excel_sheet_min_row.
+      ENDIF.
+      last_row = rows->get_max_index( ).
+      IF last_row = 0.
+        last_row = zcl_excel_common=>c_excel_sheet_max_row.
+      ENDIF.
+
```
Comment (AndreaBorgia-Abo) [inline src/zcl_excel_worksheet.clas.abap]:
```suggestion

```

Comment (AndreaBorgia-Abo) [reply src/zcl_excel_worksheet.clas.abap]: first_row and last_row are unused and the DO cycle seems to be correctly terminated via subrc = 8.

**Review by AndreaBorgia-Abo — CHANGES_REQUESTED** [review]
It works as advertised, I've tried breaking the tests in odd ways but couldn't, which is obviously good :)
If the suggested changes are accepted, then we can merge. If the reworked version makes use of those variables, please re-request a review.

**Review by AndreaBorgia-Abo — APPROVED** [review]
LGTM

**Comment by AndreaBorgia-Abo** [comment]
Rebuilding A4H, had to change cl_object_collection_iterator to zcl_excel_collection_iterator in the test program above, else it wouldn't activate anymore. Funny enough, the same goes for the other MiniSAP, NPL. Pretty sure that at the time it worked fine, I guess changes have invalidated the offending assignments.
(TLDR nothing to see here, move along)

**Pull request merged**